xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,570 @@
+    // corresponds to.
+    // A nonce becomes stale following a newer nonce being presented to the client in a
+    // DiscoveryResponse.
+    private String ldsRespNonce = "";
+    private String rdsRespNonce = "";
+
+    // Most recently requested resource name for each resource type. Note the resource_name in
+    // LDS requests will always be "xds:" URI (including port suffix if present).
+    private String rdsResourceName = "";
+
+    // A prepared ACK request. Every DiscoveryResponse is replied with a DiscoveryRequest as ACK
+    // (with version_info set to response's version_info) or NACK (with version_info set to last
+    // ACKed response's version_info).
+    @Nullable
+    private DiscoveryRequest pendingAckRequest;
+
+    private AdsStream(AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub) {
+      this.stub = checkNotNull(stub, "stub");
+    }
+
+    private void start() {
+      requestWriter = stub.withWaitForReady().streamAggregatedResources(this);
+    }
+
+    @Override
+    public void onNext(final DiscoveryResponse response) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          responseReceived = true;
+          String typeUrl = response.getTypeUrl();
+          if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+            handleLdsResponse(response);
+          } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+            handleRdsResponse(response);
+          }
+          // TODO(zdapeng): add CDS/EDS response handles.
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.fromThrowable(t).augmentDescription("ADS stream [" + this + "] had an error"));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.UNAVAILABLE.withDescription("ADS stream [" + this + "] was closed"));
+        }
+      });
+    }
+
+    private void handleStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      configWatcher.onError(error);
+      if (responseReceived || retryBackoffPolicy == null) {
+        // Reset the backoff sequence if had received a response, or backoff sequence
+        // has never been initialized.
+        retryBackoffPolicy = backoffPolicyProvider.get();
+      }
+      long delayNanos = 0;
+      if (!responseReceived) {
+        delayNanos =
+            retryBackoffPolicy.nextBackoffNanos() - stopwatch.elapsed(TimeUnit.NANOSECONDS);
+      }
+      logger.log(Level.FINE, "{0} stream closed, retry in {1} ns", new Object[]{this, delayNanos});
```

Comment:
`new Object[]{this`
You need implement `toString()` and use `InternalLogId`

Reply:
Using java's reference string is intended, this will print out something like `XdsClientImpl$AdsStream@...`. There will be at most one stream at any time, this reference can definitely represents the stream object. Printing out field details of `AdsStream` doesn't give us anything useful. We just need to know this is the `AdsStream` in `XdsClientImpl`.

Reply:
Your argument can apply to `ManagedChannelImpl`. But we were consistently using the pattern. `XdsClientImpl$AdsStream@...` looks to users that the product is hasty, although they provide similar information.